Conditionally Creating Server Spans if No Span Found in Current Context

andresbeckruiz opened the pull request:
# Description

This PR fixes issue #445. I check if if an active span is present in the current context by calling `opentelemetry.trace.get_current_span()`. If one is found, an internal span is created and the span found is used as the parent. If one is not found, a server span is created and the remote span context from the incoming request is used as a parent.

I used either this logic for all of the server instrumentations that create server spans: Falcon, Flask, Django, Pyramid, Tornado, WSGI, and ASGI. There are references to the issue that mention that FastAPI and Starlette also create server spans, but I could not find where this occurs in the opentelemetry-python-contrib repo. If someone could point out where they are created, I can add the logic I used for the other instrumentations. 

## Type of change

Please delete options that are not relevant.

- [ ✅ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
     -This feature causes server spans to be only created conditionally, so at times existing functionality will not work as expected. 


# How Has This Been Tested?

All the instrumentation tests were run and passed successfully. 

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [ ✅ ] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ✅ ] Followed the style guidelines of this project
- [ ✅ ] Changelogs have been updated
- [ ✅ ] Unit tests have been added
- [ ✅ ] Documentation has been updated

ocelotl commented:
@andresbeckruiz please make sure that all tests pass.

lonewolf3739 commented:
Two things on top of my mind that we probably should do
1) We want to continue the trace from a remote context. I think we should check that the trace id from the current span context is the same as the trace id from the incoming request before starting the INTERNAL span. Otherwise, this would lead to broken traces.
2) A check that parent span is SERVER span.

owais commented:
> 1.0 We want to continue the trace from a remote context. I think we should check that the trace id from the current span context is the same as the trace id from the incoming request before starting the INTERNAL span. Otherwise, this would lead to broken traces.

I still cannot think of any good reason why there would be locally active span that wouldn't really be part of the same trace/request so not sure how much benefit this check will provide. It's certainly not harmful but may be we can avoid extra checks on potentially hotpaths if we can't think of a edge-case to catch with this.

> 2.0 A check that parent span is SERVER span.
I see how this makes sense but can there be situations where the real server span has a child span and then this logic is triggered? Let's say if WSGI is being used with Django and then a Django middleware embeds yet another router/server library that is also instrumented. The 3rd span would create another SERVER span in that case. May be we can keep it simple and just use the local span if present without any other conditions? We can easily add more checks later if we discover any cases where this doesn't work.

lonewolf3739 commented:
Ah I see how this can get complicated quickly if we consider about all cases. Yes, maybe we can just keep it simple and revisit later if people run into these edge case scenarios.

owais commented on util/opentelemetry-util-tracer/setup.cfg:
```diff
@@ -0,0 +1,51 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Not so sure if we need a new package for this. I think we can add this to `opentelemetry-instrumentation` package given this is supposed to be used by instrumentations mainly.

andresbeckruiz replied:
I am thinking of removing this package and API, see my comment down below.

owais commented on instrumentation/opentelemetry-instrumentation-tornado/src/opentelemetry/instrumentation/tornado/__init__.py:
```diff
@@ -255,11 +256,13 @@ def _get_operation_name(handler, request):
     token = context.attach(extract(handler.request.headers))
 
-    span = tracer.start_span(
+    server_span_tracer = ServerSpanTracer(
```
This API means we extract the context every time irrespective of whether a parent span is present or not. I know it might be relatively cheap but for a service that can serve thousands of requests a second, unnecessary attribute lookups and garbage generation can quickly add up to affect performance. 

I'd change so that we first look for current span and only extract from carrier if we don't find one. 

Also, this creates one instance of ServerSpanTracer per request. Otel API was specifically designed in a way to minimize allocations by re-using things like tracers, context getters/setters, etc. We should follow that pattern and not generate new objects unnecessarily.

andresbeckruiz replied:
That makes sense; the only reason I created the ServerSpanTracer API was so that I would be able to reuse the logic of checking for a span in the current context just by calling a single `start_span()` method. I think it might be easier just to get rid of this package and API and instead use the logic directly in all the instrumentations that create server spans. If I keep using the API, I will have to add more logic for only conditionally extracting context.

owais replied:
makes sense to me. you can try having some helper functions to reuse code a bit if possible but not a requirement.

owais commented on instrumentation/opentelemetry-instrumentation-tornado/src/opentelemetry/instrumentation/tornado/__init__.py:
```diff
@@ -255,11 +256,13 @@ def _get_operation_name(handler, request):
 def _start_span(tracer, handler, start_time) -> _TraceContext:
     token = context.attach(extract(handler.request.headers))
```
wouldn't this always extract tracing information from carrier and set it in the active context resulting in the remote context always being the parent?

andresbeckruiz replied:
I did not realize that would happen. I will make edits so that I only extract context from the carrier if a span is not present in the current context.

andresbeckruiz commented:
@owais I have updated each instrumentation file and addressed your comments.

andresbeckruiz commented:
@ocelotl all tests are passing now for files that I have edited. For some reason, the psycopg2 instrumentaiton tests, util-http, lint, and docker tests are failing still, but I don't believe it's because of my edits.

owais commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -146,15 +152,27 @@ def process_request(self, request):
+                kind=SpanKind.INTERNAL,
+                start_time=request_meta.get(
+                    "opentelemetry-instrumentor-django.starttime_key"
+                ),
```
blocks like these could be turned into a reusable function

andresbeckruiz replied:
I did not end up turning these blocks into a helper method; as I mentioned in an earlier comment, different instrumentations require different attributes and have separate logic for extracting and activating context. I thought my helper method would be unnecessarily complex if I took into account all these differences, so I just implemented the logic directly in each instrumentation.

owais commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -146,15 +152,27 @@ def process_request(self, request):
-        token = attach(extract(request_meta, getter=wsgi_getter))
+        context = extract(request_meta, getter=wsgi_getter)
+
+        token = attach(context)
```
I'm a bit confused by this. Aren't we still activating the extracted span here and wouldn't `get_current_span()` call below always return that activated span? 

What we need to do this to call `get_current_span()` first and if it is invalid, then try to extract a span from the carrier and activate it. After that we just call tracer.start_span either with INTERNAL or SERVER kind. We'd probably also need to update the detach calls to do it conditionally.

andresbeckruiz commented:
@owais I have addressed your most recent comments.

owais commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -201,13 +201,22 @@ async def __call__(self, scope, receive, send):
         try:
-            with self.tracer.start_as_current_span(
-                span_name, kind=trace.SpanKind.SERVER,
-            ) as span:
```
I think the following would be cleaner:

```python
            ctx = span_kind =  None
            if trace.get_current_span() is trace.INVALID_SPAN:
                ctx = extract(scope, getter=asgi_getter)
                token = context.attach(ctx)
                span_kind = trace.SpanKind.SERVER

            with self.tracer.start_as_current_span(
                span_name, ctx, kind=span_kind
            ) as span:
```

owais commented:
This contains commits from unrelated branches. Can you please rebase and perhaps squash the commits (just this time) so it is easier to review?

andresbeckruiz commented:
@owais sorry about that, I created a new local branch to rebase and clean everything up and pushed it onto this branch. I also addressed your most recent comments.

owais commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -244,4 +249,5 @@ async def wrapped_send(message):
         finally:
-            context.detach(token)
+            if ctx is not None:
+                context.detach(token)
```
This works now but it looks like something that could easily break in future with the value of `ctx` determining whether `token` is defined or not. Why not have a none-check for `token` directly instead?

owais reviewed (commented):
Took a quick look at just one instrumentation and looks fine to me. Just concerned about code fragility a bit and left a comment about that. Will review in detail later today.

owais commented on instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py:
```diff
@@ -302,7 +311,8 @@ def _end_span_after_iterating(iterable, span, tracer, token):
             close()
         span.end()
-        context.detach(token)
+        if token is not None:
```
I think we should check for `token` not being None in every instrumentation. I see some check for `ctx` but then run operations on `token`.

owais commented on instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/callbacks.py:
```diff
@@ -178,7 +183,9 @@ def trace_tween(request):
-                context.detach(request.environ.get(_ENVIRON_TOKEN))
+                request_environ = request.environ
+                if _ENVIRON_TOKEN in request_environ.keys():
+                    context.detach(request.environ.get(_ENVIRON_TOKEN))
```
I'd just run `.get(_ENVIRON_TOKEN, None)` directly and check if the result is `None` or not before detaching. I think it'd be one less lookup.

owais reviewed (commented):
Mostly looks good. Left some minor comments. 

Major remaining blocker are tests. We need to ensure all known cases work as expected for all instrumentation that were touched. Can you please add tests for such cases?

andresbeckruiz commented:
Hi @owais, I addressed the minor comments you left. I am stuck figuring out how to implement tests for the logic I have implemented because I am unsure how to access the spans being created. For example, in the Django middleware tests I cannot find any that call the `process_request` method that I have touched. Furthermore, this method does not return the span, context, or token, which I have all updated logic for.  Therefore, I am not sure how to check for instance if the span created is of `SpanKind.INTERNAL`. Would you be able to give an example or point me in the right direction on how I could implement tests for the logic I have updated?

owais commented:
I think existing tests should continue to work with the changes you made. You need to add new tests where a span is already present in the current context before the framework receives an HTTP request. You can do this manually but it'll probably differ across frameworks. It might be simpler to have WSGI instrumentation active as well in addition to the framework instrumentation for such tests as that'd automatically reproduce the scenario. Once that is done, your test for checking spans would look exactly the same as  any other test.

owais commented:
@andresbeckruiz Thanks for contributing this to the project. It's a very valuable feature and I really appreciate all the effort you put into this. Could you please clarify if you are planning to come back to this or if we should re-assign it to someone else so they can build on top of your work?

andresbeckruiz commented:
Hi @owais sorry I did not clarify; I am pretty busy right now so I will not be coming back to this issue. I would appreciate if you  could re-assign the issue to someone else so that they can build on top of my work, thank you!
